mlflow/entities/conversion_utils.py

```diff
@@ -0,0 +1,117 @@
+try:
+    import numpy as np
+    __import_np = True
+except ImportError:
+    __import_np = False
+
+try:
+    import pyspark.ml.param
+    __import_pyspark_ml_param = True
+except ImportError:
+    __import_pyspark_ml_param = False
+
+
+try:
+    import tensorflow as tf
+    __import_np = True
+except ImportError:
+    __import_tf = False
+
+try:
+    import torch
+    __import_torch = True
+except ImportError:
+    __import_torch = False
+
+def convert_to_str_if_possible(x) -> str:
+    if type(x) == str:
+        return x
+    
+    converter_fns_to_try = [convert_to_str_if_pyspark_mlparam, convert_to_str_if_ndarray, convert_to_str_if_tensorflow_tensor, convert_to_str_if_torch_tensor]
+
+    for converter_fn in converter_fns_to_try:
+        possible_str = converter_fn(x)
+        if type(possible_str) == str:
+            return possible_str
+
+    return str(x)
+
+def convert_to_float_if_possible(x) -> float:
```

Comment:
+1, will do.

I think we need to also introduce a `MetricKey` and `MetricValue` type, so that:

```py
MetricKey = Union[str, pyspark.ml.param.Param]
MetricValue = Union[float, np.ndarray, tf.Tensor, torch.Tensor]
```

... but I am not sure how to do that without eagerly importing all these modules (which I guess we want to avoid)

Would just adding a doc, naming and tests should be sufficient?

Reply:
Don't need to add type hint for now. Avoid importing unnecessary modules is more important. The type hint requires importing modules in top level it will cause the issue.